src/compiler/checker.ts

```diff
@@ -1191,16 +1191,22 @@ namespace ts {
                                     : false;
                             }
                             if (meaning & SymbolFlags.Value && result.flags & SymbolFlags.FunctionScopedVariable) {
-                                // parameters are visible only inside function body, parameter list and return type
-                                // technically for parameter list case here we might mix parameters and variables declared in function,
-                                // however it is detected separately when checking initializers of parameters
-                                // to make sure that they reference no variables declared after them.
-                                useResult =
+                                // parameter initializer will lookup as normal variable scope when targeting es2015+
+                                if (compilerOptions.target && compilerOptions.target >= ScriptTarget.ES2015 && isParameter(lastLocation) && lastLocation.initializer === originalLocation && result.valueDeclaration !== lastLocation) {

```

Comment:
I didn't mention this in the original issue, but that's not limited to parameter intializers. It's basically every expression that could occur in a ParameterDeclaration. For example a ComputedPropertyName:

```ts
let foo = 0;

(function test({[foo]: bar}: any[]) {
    let foo = 1;
    console.log(bar); // logs 'a'
})(['a', 'b'])
```

Reply:
This is too desperate🤷‍♂️